### PR TITLE
Add support for user-supplied executors

### PR DESCRIPTION
**Pull request by stefnotch**
This lets the user supply their own async executor. This should cover advanced use-cases, such as letting the user support the async_std crate.

### CODE REVIEW

**Comment by benwis** [comment]
@stefnotch It looks like you have three PRs, #3989 #3990 and this one, that overlap and add mostly same code. Is that correct? This one adds user supplied executors, which seems to be unique. Would I want to merge this one and close the other two?

**Comment by stefnotch** [comment]
@benwis Yes indeed. I first fixed the issue with an existing executor, and then used the slightly bigger API surface to support custom executors.

So this PR builds on top of the code of #3089 .

I figured I'd submit that as separate PRs, so that if this has a design issue, then it's at least easy to just merge #3089 .

Regarding #3090 , feel free to just close it. It's significantly less important to have support for it if there's already an easy way for users to supply their own executor.

**Comment by benwis** [comment]
> @benwis Yes indeed. I first fixed the issue with an existing executor, and then used the slightly bigger API surface to support custom executors.
> 
> So this PR builds on top of the code of #3089 .
> 
> I figured I'd submit that as separate PRs, so that if this has a design issue, then it's at least easy to just merge #3089 .
> 
> Regarding #3090 , feel free to just close it. It's significantly less important to have support for it if there's already an easy way for users to supply their own executor.
> 
> (So yes, merging this and closing the other two would be perfect.)

Looks like the CI is complaining about formatting. Can you run it through cargo fmt?

**Comment by stefnotch** [comment]
Ah, I see, a nightly Rust toolchain was required to get the desired formatting output. Just as documented in the contributing.md file. Done :)

**Comment by stefnotch** [comment]
The CI is still failing. :( And while I fully understand what went wrong, I have no idea how to fix it.

What went wrong? 
`Executor::init_futures_executor()` may only be called once. If an executor has already been initialized, it panics.
It also initializes a OnceLock `static SPAWN: OnceLock<fn(PinnedFuture<()>)> = OnceLock::new();`.
Those are shared across Rust unit tests, because all tests run in the same process.

I'm not sure how to go about changing the code to make it testable.

**Comment by raskyld** [comment]
Hello!

I am pretty much interested by your PR since I need in https://github.com/leptos-rs/leptos/pull/3063, a way to provide a Custom runtime for the `wasm32-wasip2` target. Thanks for your work :) 🙏 

I think you can solve the issue by adding a const https://doc.rust-lang.org/std/sync/struct.Once.html in your `mod tests`,
then, you could add an helper function that use https://doc.rust-lang.org/std/sync/struct.Once.html#method.call_once to call the init function only once and invoke this helper in the first lines of all of your tests.

**Comment by stefnotch** [comment]
@raskyld That would work if I had one type of executor. With the custom executors part, I now have two separate executors that I have tests for
- `Executor::init_futures_executor()`
- `Executor::init_custom_executor(CustomFutureExecutor)`

I also considered writing a custom executor that lets me swap out the underlying executor, but then the question arises: How do I test that? It's a bit of nontrivial code after all.

**Comment by stefnotch** [comment]
Figured it out! I just had to use integration tests instead of unit tests.
Apparently if one puts their tests into a separate directory, then they get run in a separate process. Which solves the issue above 🎉

**Comment by stefnotch** [comment]
Question regarding the CI: Why does [CI (any_spawner)](https://github.com/leptos-rs/leptos/actions/runs/11325681448/job/31493101095?pr=3091#logs) run `cargo fmt -- --check`? I would have expected that to happen in a separate CI job that checks the entire repository, instead of every single CI job re-checking the formatting.

**Comment by raskyld** [comment]
Yeah reading the code I now see the problem, we need the tests to run as two separated process.
I don't know of any way to achieve that expect running two test command (which is kind of what you propose with the feature flag)

**Comment by raskyld** [comment]
Just saw your comments, for some reason my GH page didn't get updated 🤣 

Alright, swapping to integration tests make each module its own binary right?

**Comment by stefnotch** [comment]
> 
> Alright, swapping to integration tests make each module its own binary right?

Yes, exactly. Rust by example confirms this
> Each Rust source file in the tests directory is compiled as a separate crate.
> https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html

File: any_spawner/src/lib.rs [hunk 1]
```diff
@@ -32,8 +32,10 @@
 use std::{future::Future, pin::Pin, sync::OnceLock};
 use thiserror::Error;
 
-pub(crate) type PinnedFuture<T> = Pin<Box<dyn Future<Output = T> + Send>>;
-pub(crate) type PinnedLocalFuture<T> = Pin<Box<dyn Future<Output = T>>>;
+/// A future that has been pinned.
+pub type PinnedFuture<T> = Pin<Box<dyn Future<Output = T> + Send>>;
+/// A future that has been pinned.
```
[the remark this answers is not in the document]
Comment (raskyld) [reply any_spawner/src/lib.rs:37]: FWIW, I've commented those in my PR but idk if those are "better" tbh, the type is pretty trivial:  https://github.com/leptos-rs/leptos/pull/3063/files#diff-98c8e843d6f327eb7c721d3d6421ca2fd671e085dffeca8bed55c70ea401adc5

[the remark this answers is not in the document]
Comment (stefnotch) [reply any_spawner/src/lib.rs:37]: Sweet, then I'll leave my subpar documentation as-is, and will look forward to your PR getting merged.

(I personally am very interested in wasi as an alternative to wasm-bindgen. Currently, fully setting up wasm-bindgen with Typescript generation involves more work than necessary.)

[the remark this answers is not in the document]
Comment (raskyld) [reply any_spawner/src/lib.rs:37]: At the moment, I am focused on WASI for the Server-Side, I don't know of any initiative to bring the component model in the browser, although I think that would be possible. The browsers' engine would need to expose a world (e.g., `browser`) that exposes the various Web API to a guest component. I am curious if that's a think 🤔

**Comment by benwis** [comment]
@stefnotch got another conflict for ya

**Comment by stefnotch** [comment]
@benwis Fixed that for you :)

File: any_spawner/tests/futures_runtime.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    let _ = Executor::init_futures_executor();
+    let counter = Arc::new(AtomicUsize::new(0));
+    Executor::spawn({
+        let counter = Arc::clone(&counter);
+        async move {
+            assert_eq!(counter.fetch_add(1, Ordering::AcqRel), 0);
+        }
+    });
+    futures::executor::block_on(Executor::tick());
```
Comment (stefnotch) [inline any_spawner/tests/futures_runtime.rs:34]: I guess the fact that this test worked at all was a fluke. There's nothing in futures-executor that guarantees that one future resolves before another.

Time to either remove the test, or rewrite it to work 100% of the time.

**Pull request merged**